### DISCUSSION TITLE
posts/func-color

### DISCUSSION THREAD

**Discussion by giscus** [whexy/whexy-blog-comments#15] [General]
# posts/func-color

2015 年，著名程序员 Bob Nystrom 在博客上发表文章 What Color is Your Function，振聋发聩，发人深省，引发了行业内的激烈讨论。6 年过去了，它在各大编程语言的论坛上历久弥新。

https://www.whexy.com/posts/func-color

**Comment by it114** [comment]
类似JavaScript这种事件循环模型；如果访问网络，那就需要单独开启线程或者进程了；而宣传的一直是单线程模型

**Reply by whexy** [reply]
没错。事件循环模型在完成 I/O 任务时有可能会新建线程。

NodeJS 的内部用了 libuv，有一个线程池，通常会运行 4 个线程。这里我选一段 libuv 的文档，解释为什么线程是必要的。

>  线程会在内部使用，用来在执行系统调用时伪造异步的假象。libuv通过线程还可以使得程序异步地执行一个阻塞的任务。方法就是大量地生成新线程，然后收集线程执行返回的结果。 ——《libuv 中文教程》

我不知道访问网络时 NodeJS 底层会不会用线程去完成。如果是，那说明 NodeJS 使用*阻塞的系统调用*访问网络。但我觉得不太可能。因为 socket 的系统调用大部分是支持异步的。

事件循环每遇到一次 await 就会检查有没有 I/O 结束。在访问网络的具体例子里，每碰到一个 await，事件循环就会调用 `socket.recvfrom`，检查一下返回值是否为空。如果不为空说明 I/O 已完成，就会执行对应的回调。

无论如何，开发者是不需要自己去新建线程的。

顺带一提，不久之前 NodeJS 引入了 worker_threads，在多个线程上运行事件循环，每个线程运行一个循环。高效利用多核处理器。

**Comment by alswl** [comment]
同步异步函数问题，以前我写 tornado 时候，帮助我逃出了 callback hell。
最近几年在前端也经常被 async 关键词烦恼。

有时候知道不应该这么做，是错的，但是还是无法逃离自然的顺序思考逻辑。

传染性是「关注点」的传染。timeout / loop / spin 只是掩饰了这些传染。

**Comment by chenzx** [comment]
async/await是很优雅的东西，我把它称为“语言原生的异步IO语法”，而不仅仅只是一个语法糖，在分布式异构多核融合计算时代，“算力调度”是一个时髦的概念术语，async/await可以用命令式的外观来描述声明式的数据计算依赖（异步IO意味着代码是推迟执行的），而不是一个所谓的函数颜色问题。

让我们不要把一个程序设想为只有main一个进程启动入口，而是可以是一个分布式的协调工作的整体应用，这样的话，其实原始的非async/await的代码可以设想为一个不被异步IO机制打断的整体，可以直接包装到一个compute chunk里面去。async/await程序也不再是一个语法糖，被映射到一个事件循环或者多线程池实现机制上，而是可以设想为通过一个智能的编译转换映射到分布式多核异构环境同时执行。

说起async/await，我会想起Erlang语言（及其actor模型，源自Hoare的CSP模型？），Erlang说是当初运行在Prolog语言之上，我想Prolog不是一个逻辑约束求解语言吗？分布式系统大师Leslie Lamport最近出了一本A Science of Concurrent Programs，可以关注一下。

**Comment by boholder** [comment]
说个偏题的内容。最近一个需求（类似Grafana仪表板那种），初版实现是在单个浏览器页面内，前端会在短时间向后端发送大量网络请求查询数据，然后把数据渲染到页面上。
前端同事说这个实现会把页面卡死，我好奇问为什么，不是取数这种I/O全是Promise吗？他说在浏览器中的JS代码受限于单线程的（即使浏览器本身[为了防跨站攻击给每个site都创建了process](https://stackoverflow.com/a/17169863/11397457)）事件循环模型里，光那些渲染逻辑就会把这可怜的单线程占满，页面不会对用户操作做出反应（因为事件循环模型没空处理）。
说实话我不清楚这个解释对不对，姑且分享一下。。
